File: sandbox/src/streaming/sse.ts
import type { ExecStreamFrame } from '../types';

function parseSseBlock(block: string): ExecStreamFrame | null {
  const dataLine = block
    .split('\n')
    .map((line) => line.trim())
    .find((line) => line.startsWith('data:'));

  if (!dataLine) {
    return null;
  }

  const json = dataLine.slice('data:'.length).trim();
  if (!json) {
    return null;
  }

  return JSON.parse(json) as ExecStreamFrame;
}

/**
 * Incrementally parse SSE (`text/event-stream`) frames from a byte stream.
 * Comment lines (`: open`, `: ping`) are ignored.
 */
export async function* parseSseFrames(
  body: ReadableStream<Uint8Array>,
  signal?: AbortSignal,
): AsyncGenerator<ExecStreamFrame> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  const abort = () => {
    reader.cancel().catch(() => undefined);
  };

  signal?.addEventListener('abort', abort, { once: true });

  try {
    while (true) {
      if (signal?.aborted) {
        throw signal.reason ?? new Error('Stream aborted');
      }

      const { value, done } = await reader.read();
      if (done) {
        break;
      }

      if (value) {
        buffer += decoder.decode(value, { stream: true });
      }

      let boundary = buffer.indexOf('\n\n');
      while (boundary !== -1) {
        const block = buffer.slice(0, boundary);
        buffer = buffer.slice(boundary + 2);

        const frame = parseSseBlock(block);
        if (frame) {
          yield frame;
        }

        boundary = buffer.indexOf('\n\n');
      }
    }

    buffer += decoder.decode();
    if (buffer.trim().length > 0) {
      const frame = parseSseBlock(buffer);
      if (frame) {
        yield frame;
      }
    }
  } finally {
    signal?.removeEventListener('abort', abort);
    reader.releaseLock();
  }
}

File: sandbox-python/brimble_sandbox/streaming/sse.py
"""Incremental SSE (`text/event-stream`) frame parser."""

from __future__ import annotations

import json
from collections.abc import Iterator
from typing import IO, cast

from ..types import ExecStreamFrame


def _parse_sse_block(block: str) -> ExecStreamFrame | None:
    data_line = next((line.strip() for line in block.split("\n") if line.strip().startswith("data:")), None)
    if not data_line:
        return None

    payload = data_line[len("data:") :].strip()
    if not payload:
        return None

    return cast(ExecStreamFrame, json.loads(payload))


def parse_sse_frames(body: IO[bytes]) -> Iterator[ExecStreamFrame]:
    """Parse SSE frames from a byte stream. Comment blocks (`: open`, `: ping`) are ignored."""
    buffer = ""

    while True:
        chunk = body.read(4096)
        if not chunk:
            break

        buffer += chunk.decode("utf-8")

        while "\n\n" in buffer:
            block, buffer = buffer.split("\n\n", 1)
            frame = _parse_sse_block(block)
            if frame is not None:
                yield frame

    if buffer.strip():
        frame = _parse_sse_block(buffer)
        if frame is not None:
            yield frame
